RIOT-OS/RIOT
# netif: introduce generic network interface descriptor

### Contribution description
This PR changes the way a `netif_t` object is represented. Currently the representation is stack specific (e.g https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/netif/include/netif_types.h#L29). With this PR, an interface is always represented by a `netif_t` object (where stack specific network interfaces can inherit).

The consequences of these changes are:
- It's possible to move the stack dependent iteration logic to generic `netif_t` functions (see [here](https://github.com/RIOT-OS/RIOT/compare/master...jia200x:pr/netif_new_interface?expand=1#diff-bb8db03c5c9902b4f29d0c1356ee77d3R35))
- It makes easier to add stack independent functions send and receive data from an interface (e.g `netif_send(netif_t *netif)`, `netif_recv(netif_t *netif)`). ~TBD!~ (see #11879)

The `netif_t` is now a linked list that point to the next interface, so it's required to register netifs via the newly introduced `netif_register` function.
Also, the NETIF_INVALID macro is not needed anymore (an invalid interface is mapped to NULL).

This PR is needed in order to have MAC/PHY layers independent of the network stack.





### Testing procedure
It should be enough to run `tests/gnrc_netif` test.
Check if the documentation builds without issues: `make doc`




### Issues/PRs references
It somehow relates to #11483.
